packages/@react-spectrum/cards/src/GridLayout.tsx

```diff
@@ -0,0 +1,283 @@
+    let width = availableWidth - (this.minSpace.width * Math.max(0, this.numColumns - 1));
+
+    // Compute the item width based on the space available
+    let itemWidth = Math.floor(width / this.numColumns);
+    itemWidth = Math.max(this.minItemSize.width, Math.min(this.maxItemSize.width, itemWidth));
+
+    // Compute the item height, which is proportional to the item width
+    let t = ((itemWidth - this.minItemSize.width) / this.minItemSize.width);
+    let itemHeight = this.minItemSize.height + this.minItemSize.height * t;
+    itemHeight = Math.max(this.minItemSize.height, Math.min(this.maxItemSize.height, itemHeight)) + this.itemPadding;
+
+    this.itemSize = new Size(itemWidth, itemHeight);
+
+    // Compute the horizontal spacing and content height
+    this.horizontalSpacing = Math.floor((visibleWidth - this.numColumns * this.itemSize.width) / (this.numColumns + 1));
+
+    let y = this.margin;
+    let index = 0;
+    for (let node of this.collection) {
+      let layoutInfo = this.buildChild(node, y, index);
+      y = layoutInfo.rect.maxY;
+      index++;
+    }
+
+    if (this.isLoading) {
+      let loaderY = y;
+      let loaderHeight = 60;
+      // If there aren't any items, make loader take all avaliable room and remove margin from y calculation
+      // so it doesn't scroll
+      if (this.collection.size === 0) {
+        loaderY = 0;
+        loaderHeight = visibleHeight || 60;
+      }
+
+      let rect = new Rect(0, loaderY, visibleWidth, loaderHeight);
+      let loader = new LayoutInfo('loader', 'loader', rect);
+      this.layoutInfos.set('loader', loader);
+      y = loader.rect.maxY;
+    }
+
+    if (this.collection.size === 0 && !this.isLoading) {
+      let rect = new Rect(0, 0, visibleWidth, visibleHeight);
+      let placeholder = new LayoutInfo('placeholder', 'placeholder', rect);
+      this.layoutInfos.set('placeholder', placeholder);
+      y = placeholder.rect.maxY;
+    }
+
+    this.contentSize = new Size(visibleWidth, y);
+  }
+
+  buildChild(node: Node<T>, y: number, index: number): LayoutInfo {
+    let row = Math.floor(index / this.numColumns);
+    let column = index % this.numColumns;
+    let x = this.margin + column * (this.itemSize.width + this.horizontalSpacing);
+    y = this.margin + row * (this.itemSize.height + this.minSpace.height);
+
+    let rect = new Rect(x, y, this.itemSize.width, this.itemSize.height);
+    let layoutInfo = new LayoutInfo(node.type, node.key, rect);
+    this.layoutInfos.set(node.key, layoutInfo);
+    return layoutInfo;
+  }
+
+  // Since the collection doesn't represent the visual layout, need to calculate what row and column the current key is in,
+  // then return the key that occupies the row + column below. This can be done by figuring out how many cards exist per column then dividing the
+  // collection contents by that number (which will give us the row distribution)
+  getKeyBelow(key: Key) {
```

Comment:
not sure what you mean, left/right is in the base layout and translates to left = previous and right = next (not necessarily spacial left and right since it should advance to the next row if you hit ArrowRight on the last item in the previous row)

Reply:
Right, but this is a grid layout, so left/right will always be findIndex(key) + 1 or findIndex(key) - 1 , we wouldn't want to do a closest on them?